clamp/clamp.tex

```diff
@@ -0,0 +1,331 @@
+
+\begin{figure}[ht]
+    \includegraphics[width=16cm]{clamp_rust.png}
+    \centering
+    \label{fig:code}
+\end{figure}
+
+% Change: go back to the previous Git push.
+% TODO: add clamp sensitivity?
+
+\subsection{Pseudocode in Python}\label{sec:pseudocode}
+We present a simplified Python-like pseudocode of the Rust implementation below. The necessary definitions for the pseudocode can be found at \href{https://www.overleaf.com/project/60d215bf90b337ac02200a99}{``List of definitions used in the pseudocode"}. 
+
+%\silvia{We could generalize the input domain below from float to a any general (unspecified) type that admits total ordering, and then add the corresponding precondition to assert that the type \texttt{T} for $L$ and $U$ has the trait \texttt{TotalOrd}. Note that partial ordering is not enough, and we might not be able to clamp in the domain. However, TotalOrd is still not implemented in the OpenDP library.}
+
+%\newpage
+
+\subsubsection*{Preconditions}
+To ensure the correctness of the output, we require the following preconditions:
+
+\begin{itemize}
+    \item \textbf{User-specified types:}
+    \begin{itemize}
+        \item Type \texttt{T} must have trait \texttt{TotalOrd}.\footnote{For now, the OpenDP library only implements \texttt{PartialOrd}, but \texttt{TotalOrd} will soon be implemented.}
+    \end{itemize}
+    \item \texttt{input\_domain:} any vector of elements of type \texttt{T}.
+    \item \texttt{output\_domain:} any vector of elements of type \texttt{T} which are contained in the interval \texttt{[L, U]}, where 
+    \texttt{L} and \texttt{U} are of type \texttt{T}. \silvia{Problem: \texttt{L} and \texttt{U} appear only later in the pseudocode, as discussed on July 13.}
+    \item \texttt{input\_metric:} symmetric distance.
+    \item \texttt{output\_metric:} symmetric distance.
+\end{itemize}
+
+\subsubsection*{Postconditions}
+\begin{itemize}
+    \item A \texttt{Transformation} is returned (i.e., if a \texttt{Transformation} cannot be returned successfully, then an error should be returned).
+\end{itemize}
+
+% Perhaps: change L for lower and U for upper
+\begin{lstlisting}[language=Python, escapechar=|] 
+def MakeClamp(L: T, U: T): |\label{line:def}|
+    if L > U: raise Exception('Invalid parameters') |\label{line:excep}|
+    def Relation(d_in: u32, d_out: u32) -> bool: |\label{line:rel}|
+        return d_out >= d_in*1
+    
+    def function(data: Vec(T)) -> Vec(T): |\label{line:fn}|
+        def clamp(x: T) -> T: |\label{line:clamp}|
+            return max(min(x, U), L)
+        return list(map(clamp, data)) |\label{line:map}|
```

Comment:
where can I find definitions of max, min, list, map?  Are these built-in Rust functions?

Reply:
Max and min are defined in the pseudocode definitions file (see Definitions 4.17 and 4.18). Map and list are built-in functions, but I can add them to the pseudocode definitions file. In the case of clamp, the Rust code does not directly use a list, but talking to Mike we settled on that the best Python-Rust blend term for this line would be to use a list, instead of the Rust .collect() function that appears in the code.

Reply:
I do think it would be good to add map and state in the pseudocode defs file that it is representing the Rust function collect. Actually doesn't map by definition return a list?  So what is the list(.) function doing?

Reply:
nvm, I see my questions are answered below- but the fact that Grace and I had the same question reinforces that it would be good to add defs of map and list into the pseudocode_defns file

Reply:
Yes, they are added now